File: src/main/java/me/kosik/interwalled/algorithm/ailist/AIList.java
package me.kosik.interwalled.algorithm.ailist;

import me.kosik.interwalled.algorithm.Interval;
import me.kosik.interwalled.algorithm.IntervalHolder;
import me.kosik.interwalled.algorithm.OverlapIterator;

import java.util.ArrayList;
import java.util.Iterator;

// Augmented Interval List implementation
public class AIList<T> implements IntervalHolder<T> {

    // All intervals
    private final ArrayList<Interval<T>> intervals;

    // Number of components (sub lists of intervals).
    private final int componentsCount;

    // Mapping of component index to component's length.
    private final ArrayList<Integer> componentsLengths;

    // Mapping of component index to component's starting index (offset) in {intervals}.
    private final ArrayList<Integer> componentsStartIndexes;

    // Mapping of component index to maximum 'end' value in of all component's intervals.
    private final ArrayList<Long> componentsMaxEnds;

    AIList(
            final ArrayList<Interval<T>> intervals,
            final int componentsCount,
            final ArrayList<Integer> componentsLengths,
            final ArrayList<Integer> componentsStartIndexes,
            final ArrayList<Long> componentsMaxEnds)
    {
        this.intervals = intervals;
        this.componentsCount = componentsCount;
        this.componentsLengths = componentsLengths;
        this.componentsStartIndexes = componentsStartIndexes;
        this.componentsMaxEnds = componentsMaxEnds;
    }

    @Override
    public OverlapIterator<T> overlapping(Interval<T> interval) {
        return new AIListIterator<T>(interval.start(), interval.end(), this);
    }

    /* OverlapIterator interface. */

    int size() {
        return intervals.size();
    }

    int getComponentStartIndex(final int componentIndex) {
        return componentsStartIndexes.get(componentIndex);
    }

    int getComponentLength(final int componentIndex) {
        return componentsLengths.get(componentIndex);
    }

    long getComponentMaxEnd(final int componentIndex) {
        int componentStartIndex = getComponentStartIndex(componentIndex);
        int componentLength = getComponentLength(componentIndex);
        int componentEndIndex = componentStartIndex + componentLength - 1;

        return componentsMaxEnds.get(componentEndIndex);
    }

    long getIntervalMaxEnd(final int intervalIndex) {
        return componentsMaxEnds.get(intervalIndex);
    }

    int getComponentsCount() {
        return componentsCount;
    }

    Interval<T> getInterval(final int index) {
        return intervals.get(index);
    }

    ArrayList<Interval<T>> getIntervals() {
        return intervals;
    }
}

File: ailist/src/main/scala/me/kosik/interwalled/ailist/AIList.scala
package me.kosik.interwalled.ailist

import me.kosik.interwalled.ailist.model.{Interval, Intervals}


case class AIList(
  intervals: Intervals,
  componentsCount: Int,
  componentsLengths: Array[Int],
  componentsStartIndexes: Array[Int],
  componentsMaxEnds: Array[Long]
) {
  def overlapping(interval: Interval) =
    new AIListIterator(interval.from, interval.to, this)

  def length: Int =
    intervals.length

  override def toString: String = {
    val componentsStr = componentsLengths.mkString(", ")
    f"AIList { Components: $componentsCount { $componentsStr }}"
  }

  /* OverlapIterator interface. */
  // FIXME: Remove those methods, use immutable fields instead.

  private[ailist] def getComponentStartIndex(componentIndex: Int) =
    componentsStartIndexes(componentIndex)

  private[ailist] def getComponentLength(componentIndex: Int) =
    componentsLengths(componentIndex)

  private[ailist] def getComponentMaxEnd(componentIndex: Int) = {
    val componentStartIndex = getComponentStartIndex(componentIndex)
    val componentLength = getComponentLength(componentIndex)
    val componentEndIndex = componentStartIndex + componentLength - 1
    componentsMaxEnds(componentEndIndex)
  }

  private[ailist] def getIntervalMaxEnd(intervalIndex: Int) =
    componentsMaxEnds(intervalIndex)

  private[ailist] def getComponentsCount =
    componentsCount

  private[ailist] def getInterval(index: Int) =
    intervals.get(index)

  private[ailist] def getIntervals =
    intervals
}


